# CS4044D - Machine Learning Assignment
- > **Submitted By - Vishnu Sajith, B180474CS**

<h3>Importing Libraries</h3

In [1]:
import numpy as np
import math

<h1>Qn1</h1>

In [2]:
def g(x,mean,cov,d,p):
    a = np.transpose(x-mean)
    if d == 1:
        cov_inverse = 1/cov
        det_cov = cov
    else:
        cov_inverse = np.linalg.inv(cov)
        det_cov = np.linalg.det(cov)
    a = np.dot(cov_inverse,a)
    a = np.dot((x-mean),a)
    a *= -0.5
    b = -0.5*d*np.log(2*math.pi)
    c = -0.5*np.log(det_cov)
    d = np.log(p)
    return a + b + c + d

<h2>Data Pre-processing for Q2</h2>

<h4>Sample Data</h4>

In [3]:
w1x1 = [-5.01, -5.43, 1.08, 0.86, -2.67, 4.94, -2.51, -2.25, 5.56, 1.03]
w1x2 = [-8.12, -3.48, -5.52, -3.78, 0.63, 3.29, 2.09, -2.13, 2.86, -3.33]
w1x3 = [-3.68, -3.54, 1.66, -4.11, 7.39, 2.08, -2.59, -6.94, -2.26, 4.33]

In [4]:
w2x1 = [-0.91, 1.30, -7.75, -5.47, 6.14, 3.60, 5.37, 7.18, -7.39, -7.50]
w2x2 = [-0.18, -2.06, -4.54, 0.50, 5.72, 1.26, -4.63, 1.46, 1.17, -6.32]
w2x3 = [-0.05, -3.53, -0.95, 3.92, -4.85, 4.36, -3.65, -6.66, 6.30, -0.31]

In [5]:
w3x1 = [5.35, 5.12, -1.34, 4.48, 7.11, 7.17, 5.75, 0.77, 0.90, 3.52]
w3x2 = [2.26, 3.22, -5.31, 3.42, 2.39, 4.33, 3.97, 0.27, -0.43, -0.36]
w3x3 = [8.13, -2.66, -9.87, 5.19, 9.21, -0.98, 6.65, 2.41, -8.71, 6.43]

<h4>3 Features - x1,x2,x3</h4>

##### - Mean

In [6]:
w1_3d_mean = np.array([[np.mean(w1x1),np.mean(w1x2),np.mean(w1x3)]])
w2_3d_mean = np.array([[np.mean(w2x1),np.mean(w2x2),np.mean(w2x3)]])
w3_3d_mean = np.array([[np.mean(w3x1),np.mean(w3x2),np.mean(w3x3)]])

##### - Covariance

In [7]:
w1_3d = np.stack((w1x1,w1x2,w1x3),axis = 0)
w1_3d_cov = np.cov(w1_3d)
w2_3d = np.stack((w2x1,w2x2,w2x3),axis = 0)
w2_3d_cov = np.cov(w2_3d)
w3_3d = np.stack((w3x1,w3x2,w3x3),axis = 0)
w3_3d_cov = np.cov(w3_3d)

<h4>2 Features - x1,x2</h4>

##### - Mean

In [8]:
w1_2d_mean = np.array([[np.mean(w1x1),np.mean(w1x2)]])
w2_2d_mean = np.array([[np.mean(w2x1),np.mean(w2x2)]])
w3_2d_mean = np.array([[np.mean(w3x1),np.mean(w3x2)]])

##### - Covariance

In [9]:
w1_2d = np.stack((w1x1,w1x2),axis = 0)
w1_2d_cov = np.cov(w1_2d)
w2_2d = np.stack((w2x1,w2x2),axis = 0)
w2_2d_cov = np.cov(w2_2d)
w3_2d = np.stack((w3x1,w3x2),axis = 0)
w3_2d_cov = np.cov(w3_2d)

<h4>1 Feature - x1</h4>

##### - Mean

In [10]:
w1_1d_mean = np.mean(w1x1)
w2_1d_mean = np.mean(w2x1)
w3_1d_mean = np.mean(w3x1)

##### - Covariance

In [11]:
w1_1d_cov = np.cov(w1x1)
w2_1d_cov = np.cov(w2x1)
w3_1d_cov = np.cov(w3x1)

### Sample Data

In [12]:
w1 = np.stack((w1x1,w1x2,w1x3),axis = 1)
w2 = np.stack((w2x1,w2x2,w2x3),axis = 1)
w3 = np.stack((w3x1,w3x2,w3x3),axis = 1)

# Qn2a

In [13]:
def dichotomizer(w,x,m1,m2,m3,c1,c2,c3,d):
    if g(x,m1,c1,d,0.5) > max(g(x,m2,c2,d,0.5),g(x,m3,c3,d,0)):
        print(str(w) + " is classified as W1 considering " + str(d) + " features " + str(x))
    elif g(x,m2,c2,d,0.5) > max(g(x,m1,c1,d,0.5),g(x,m3,c3,d,0)):
        print(str(w) + " is classified as W2 considering " + str(d) + " features " + str(x))
    elif g(x,m3,c3,d,0) > max(g(x,m1,c1,d,0.5),g(x,m2,c2,d,0.5)):
        print(str(w) + " is classified as W3 considering " + str(d) + " features " + str(x))

In [14]:
print("\n-------------------------------To be classified as W1-------------------------------\n")
for x in w1:
    dichotomizer(x,x[0],w1_1d_mean,w2_1d_mean,w3_1d_mean,w1_1d_cov,w2_1d_cov,w3_1d_cov,1)
print("\n-------------------------------To be classified as W2-------------------------------\n")
for x in w2:
    dichotomizer(x,x[0],w1_1d_mean,w2_1d_mean,w3_1d_mean,w1_1d_cov,w2_1d_cov,w3_1d_cov,1)
print("\n-------------------------------To be classified as W3-------------------------------\n")
for x in w3:
    dichotomizer(x,x[0],w1_1d_mean,w2_1d_mean,w3_1d_mean,w1_1d_cov,w2_1d_cov,w3_1d_cov,1)


-------------------------------To be classified as W1-------------------------------

[-5.01 -8.12 -3.68] is classified as W1 considering 1 features -5.01
[-5.43 -3.48 -3.54] is classified as W2 considering 1 features -5.43
[ 1.08 -5.52  1.66] is classified as W1 considering 1 features 1.08
[ 0.86 -3.78 -4.11] is classified as W1 considering 1 features 0.86
[-2.67  0.63  7.39] is classified as W1 considering 1 features -2.67
[4.94 3.29 2.08] is classified as W2 considering 1 features 4.94
[-2.51  2.09 -2.59] is classified as W1 considering 1 features -2.51
[-2.25 -2.13 -6.94] is classified as W1 considering 1 features -2.25
[ 5.56  2.86 -2.26] is classified as W2 considering 1 features 5.56
[ 1.03 -3.33  4.33] is classified as W1 considering 1 features 1.03

-------------------------------To be classified as W2-------------------------------

[-0.91 -0.18 -0.05] is classified as W1 considering 1 features -0.91
[ 1.3  -2.06 -3.53] is classified as W1 considering 1 features 1.3
[-7.75 -

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log
  


# Q2b

In [15]:
def missclassification(w1,w2,w3,m1,m2,m3,c1,c2,c3,d):
    count1 = 0
    for x in w1:
        if g(x,m1,c1,d,0.5) < max(g(x,m2,c2,d,0.5),g(x,m3,c3,d,0)):
            count1 += 1
    print("% of Missclassified Points in W1 is: " + str(count1*100/len(w1))+str("%"))
    count2 = 0
    for x in w2:
        if g(x,m2,c2,d,0.5) < max(g(x,m1,c1,d,0.5),g(x,m3,c3,d,0)):
            count2 += 1
    print("% of Missclassified Points in W2 is: " + str(count2*100/len(w2))+str("%"))
    count3 = 0
    for x in w3:
        if g(x,m3,c3,d,0) < max(g(x,m1,c1,d,0.5),g(x,m2,c2,d,0.5)):
            count3 += 1
    print("% of Missclassified Points in W3 is: " + str(count3*100/len(w3)) + "%")
    print("% of Missclassified Points in the dataset: " + str((count1+count2+count3)*100/(len(w1)+len(w1)+len(w1))) + "%\n")

In [16]:
print("\n----------Percentage of Missclassification----------\n")
missclassification(w1x1,w2x1,w3x1,w1_1d_mean,w2_1d_mean,w3_1d_mean,w1_1d_cov,w2_1d_cov,w3_1d_cov,1)


----------Percentage of Missclassification----------

% of Missclassified Points in W1 is: 30.0%
% of Missclassified Points in W2 is: 30.0%
% of Missclassified Points in W3 is: 100.0%
% of Missclassified Points in the dataset: 53.333333333333336%



C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log
  


# Qn2c

In [17]:
w1_2d = np.stack((w1x1,w1x2),axis = 1)
w2_2d = np.stack((w2x1,w2x2),axis = 1)
w3_2d = np.stack((w3x1,w3x2),axis = 1)
print("\n-------------------------------To be classified as W1-------------------------------\n")
for x in w1:
    dichotomizer(x,x[0:2],w1_2d_mean,w2_2d_mean,w3_2d_mean,w1_2d_cov,w2_2d_cov,w3_2d_cov,2)
print("\n-------------------------------To be classified as W2-------------------------------\n")
for x in w2:
    dichotomizer(x,x[0:2],w1_2d_mean,w2_2d_mean,w3_2d_mean,w1_2d_cov,w2_2d_cov,w3_2d_cov,2)
print("\n-------------------------------To be classified as W3-------------------------------\n")
for x in w3_2d:
    dichotomizer(x,x[0:2],w1_2d_mean,w2_2d_mean,w3_2d_mean,w1_2d_cov,w2_2d_cov,w3_2d_cov,2)
print("\n--------------------------Percentage of Missclassification--------------------------\n")
missclassification(w1_2d,w2_2d,w3_2d,w1_2d_mean,w2_2d_mean,w3_2d_mean,w1_2d_cov,w2_2d_cov,w3_2d_cov,2)


-------------------------------To be classified as W1-------------------------------

[-5.01 -8.12 -3.68] is classified as W1 considering 2 features [-5.01 -8.12]
[-5.43 -3.48 -3.54] is classified as W2 considering 2 features [-5.43 -3.48]
[ 1.08 -5.52  1.66] is classified as W1 considering 2 features [ 1.08 -5.52]
[ 0.86 -3.78 -4.11] is classified as W1 considering 2 features [ 0.86 -3.78]
[-2.67  0.63  7.39] is classified as W2 considering 2 features [-2.67  0.63]
[4.94 3.29 2.08] is classified as W2 considering 2 features [4.94 3.29]
[-2.51  2.09 -2.59] is classified as W2 considering 2 features [-2.51  2.09]
[-2.25 -2.13 -6.94] is classified as W1 considering 2 features [-2.25 -2.13]
[ 5.56  2.86 -2.26] is classified as W2 considering 2 features [5.56 2.86]
[ 1.03 -3.33  4.33] is classified as W1 considering 2 features [ 1.03 -3.33]

-------------------------------To be classified as W2-------------------------------

[-0.91 -0.18 -0.05] is classified as W1 considering 2 features 

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log
  


# Qn2d

In [18]:
print("\n-------------------------------To be classified as W1-------------------------------\n")
for x in w1:
    dichotomizer(x,x,w1_3d_mean,w2_3d_mean,w3_3d_mean,w1_3d_cov,w2_3d_cov,w3_3d_cov,3)
print("\n-------------------------------To be classified as W2-------------------------------\n")
for x in w2:
    dichotomizer(x,x,w1_3d_mean,w2_3d_mean,w3_3d_mean,w1_3d_cov,w2_3d_cov,w3_3d_cov,3)
print("\n-------------------------------To be classified as W3-------------------------------\n")
for x in w3:
    dichotomizer(x,x,w1_3d_mean,w2_3d_mean,w3_3d_mean,w1_3d_cov,w2_3d_cov,w3_3d_cov,3)
print("\n--------------------------Percentage of Missclassification--------------------------\n")
missclassification(w1,w2,w3,w1_3d_mean,w2_3d_mean,w3_3d_mean,w1_3d_cov,w2_3d_cov,w3_3d_cov,3)


-------------------------------To be classified as W1-------------------------------

[-5.01 -8.12 -3.68] is classified as W1 considering 3 features [-5.01 -8.12 -3.68]
[-5.43 -3.48 -3.54] is classified as W1 considering 3 features [-5.43 -3.48 -3.54]


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log
  


[ 1.08 -5.52  1.66] is classified as W1 considering 3 features [ 1.08 -5.52  1.66]
[ 0.86 -3.78 -4.11] is classified as W1 considering 3 features [ 0.86 -3.78 -4.11]
[-2.67  0.63  7.39] is classified as W2 considering 3 features [-2.67  0.63  7.39]
[4.94 3.29 2.08] is classified as W1 considering 3 features [4.94 3.29 2.08]
[-2.51  2.09 -2.59] is classified as W1 considering 3 features [-2.51  2.09 -2.59]
[-2.25 -2.13 -6.94] is classified as W1 considering 3 features [-2.25 -2.13 -6.94]
[ 5.56  2.86 -2.26] is classified as W2 considering 3 features [ 5.56  2.86 -2.26]
[ 1.03 -3.33  4.33] is classified as W1 considering 3 features [ 1.03 -3.33  4.33]

-------------------------------To be classified as W2-------------------------------

[-0.91 -0.18 -0.05] is classified as W2 considering 3 features [-0.91 -0.18 -0.05]
[ 1.3  -2.06 -3.53] is classified as W2 considering 3 features [ 1.3  -2.06 -3.53]
[-7.75 -4.54 -0.95] is classified as W2 considering 3 features [-7.75 -4.54 -0.95]
[-5.47

# Qn2e

- > We can see that misclassification is highest when it comes to **2-features; 63.33%.**
- > The lowest misclassifications occur in the case of **3-features; 43.33%**
- > The misclassification in the case of **1-feature is 53.33%**

# Qn2f

In [19]:
def classify(x,d):
    if d == 1:
        dichotomizer(x,x[0][0],w1_1d_mean,w2_1d_mean,w3_1d_mean,w1_1d_cov,w2_1d_cov,w3_1d_cov,1)
    elif d == 2:
        dichotomizer(x,x[0][0:2],w1_2d_mean,w2_2d_mean,w3_2d_mean,w1_2d_cov,w2_2d_cov,w3_2d_cov,2)
    elif d == 3:
        dichotomizer(x,x,w1_3d_mean,w2_3d_mean,w3_3d_mean,w1_3d_cov,w2_3d_cov,w3_3d_cov,3)

In [20]:
input_sample = [[1,2,1],[5,3,2],[0,0,0],[1,0,0]]
for sample in input_sample:
    for feature in range(1,4):
        classify([sample],feature)
    print("--------------------------------------------------------------------")

[[1, 2, 1]] is classified as W1 considering 1 features 1
[[1, 2, 1]] is classified as W1 considering 2 features [1, 2]
[[1, 2, 1]] is classified as W2 considering 3 features [[1, 2, 1]]
--------------------------------------------------------------------
[[5, 3, 2]] is classified as W2 considering 1 features 5
[[5, 3, 2]] is classified as W2 considering 2 features [5, 3]
[[5, 3, 2]] is classified as W1 considering 3 features [[5, 3, 2]]
--------------------------------------------------------------------
[[0, 0, 0]] is classified as W1 considering 1 features 0
[[0, 0, 0]] is classified as W1 considering 2 features [0, 0]
[[0, 0, 0]] is classified as W1 considering 3 features [[0, 0, 0]]
--------------------------------------------------------------------
[[1, 0, 0]] is classified as W1 considering 1 features 1
[[1, 0, 0]] is classified as W1 considering 2 features [1, 0]
[[1, 0, 0]] is classified as W1 considering 3 features [[1, 0, 0]]
-------------------------------------------------

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log
  


### Inference and Conclusion

- > From our observation above, we know that **percentage of misclassification is least when we consider 3-features**
- > Hence we can conclude that the sample [1,2,1] and [5,3,2] are misclassified as W1 and W2 respectively when we consider 1 feature and 2 features.
- > Hence, the sample **[1,2,1] most likely belongs to class W2 and sample [5,3,2] most likely belongs to W1**
- > The sample **[0,0,0] and [1,0,0] is classified as W1 considering 1,2, and 3 features.**